### ISSUE TITLE
newCommandTimeout not working if the last comman is cause exeception(like NoSuchElementError)

### ISSUE THREAD

**Issue by linyayu** [appium/appium#20368]
### Do I have the most recent component updates?

- [X] I use the most recent available driver/plugin and server versions

### Is the component officially supported by the Appium team?

- [X] I have verified the component repository is present under the Appium organization in GitHub

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

appium session no quit after noCommandTimeout

### Expected Behavior

appium session will quit after noCommandTimeout

### Minimal Reproducible Example

It happened on commad appear exception(like NoSuchElement) with plugin(image)

### Environment

- Operating system:
- Appium server version (output of `appium --version`):
- Appium driver(s) and their version(s) (`appium driver list`):
- Appium plugin(s) and their version(s) (`appium plugin list`):
- Node.js version (output of `node --version`):
- `npm` version (output of `npm --version`):
- Last component(s) version which did _not_ exhibit the problem:
- Platform and version under test:
- Real device or emulator/simulator:

### Link to Appium Logs

https://gist.github.com/linyayu/33786ee48be917d9a30e58af8e0e346a

### Further Information
It happened on commad appear exception(like NoSuchElement) with plugin(image)
check the code ,refresh noCommandTimer after  Command execute
<img width="1041" alt="Screenshot 2024-07-12 at 11 15 11" src="https://github.com/user-attachments/assets/72d0ba6d-c1ee-4e47-8c4e-f6b604b60d19">

_No response_

**Comment by mykola-mokhnach** [comment]
Please fill out component versions properly.
Also, I don't observe any timeout errors in the attached log. Is it a valid one?

**Comment by linyayu** [comment]
**Reprodece:**
1.start a session and set newCommandTimeout 
2.send last command which cause exception like (NoSuchElement )

session will keep alive forver

**Comment by mykola-mokhnach** [comment]
> uiautomator2@2.29.6 [installed (npm)]
> images@2.1.4 [installed (npm)]

This is not the latest version. Also, I still don't see what the server version is. Unfortunately we do not patch older version, so make sure the issue reproducible in the most recent version

**Comment by linyayu** [comment]
**Appium version 2.10.3**
✔ Listing available drivers
**- xcuitest@7.18.0 [installed (npm)]
- uiautomator2@3.6.1 [installed (npm)]**
✔ Listing available plugins
**- relaxed-caps@1.0.6 [installed (npm)]
- images@2.1.2 [installed (npm)]**

Still can reproduce in the most recent version
https://gist.github.com/linyayu/bad0ec39c3312c950b64b0bfb9fac27b
@mykola-mokhnach 
Maybe you can reproduce on your side,It's a aways issues

My guess is that the noCommandTimer was refreshed after the execution of the command, but an exception to the execution of the command caused the noCommandTimer not to restart
I

**Comment by mykola-mokhnach** [comment]
I still don't observe any issues in the attached log. The newCommandTimeout is only applied between two session commands. The flow is:
1. Start the timer
2. Run a session command
3. Run the next session command

If the interval between 2 and 3 is greater than the given time interval then an exception will be thrown. In the log above I can only observe 1. and 2., but 3. has never been arrived as getSessions is not a session command

**Comment by linyayu** [comment]
> I still don't observe any issues in the attached log. The newCommandTimeout is only applied between two session commands. The flow is:
> 
> 1. Start the timer
> 2. Run a session command
> 3. Run the next session command
> 
> If the interval between 2 and 3 is greater than the given time interval then an exception will be thrown. In the log above I can only observe 1. and 2., but 3. has never been arrived as getSessions is not a session command

Isn't newCommandTimeout intended to exit a session if there is no session command for a long time? Now our expectation is that 3 has not been executed, but has not exited the session

**Comment by linyayu** [comment]
if 2 is normal execute,we can found 3 never been arrived that will cause session quit after newCommandTimeout

**Comment by mykola-mokhnach** [comment]
> Isn't newCommandTimeout intended to exit a session if there is no session command for a long time? Now our expectation is that 3 has not been executed, but the session exited

No, this capability does not work like that. It only measures interval between two commands in the same session

**Comment by linyayu** [comment]
> > Isn't newCommandTimeout intended to exit a session if there is no session command for a long time? Now our expectation is that 3 has not been executed, but the session exited
> 
> No, this capability does not work like that. It only measures interval between two commands in the same session

If that ,why will have this result ?(if 2 is normal execute,we can found 3 never been arrived that will cause session quit after newCommandTimeout)

**Comment by mykola-mokhnach** [comment]
This is how it was always working before. It is the client's responsibility to properly manage sessions and close them if they are not needed anymore. There is also `session-override` mode, where the next session would override/delete the provious one if exists

**Comment by linyayu** [comment]
> This is how it was always working before. It is the client's responsibility to properly manage sessions and close them if they are not needed anymore. There is also `session-override` mode, where the next session would override/delete the provious one if exists

Look at the code implementation of appium，noCommandTimer is a timer task which don't need 3 to trigger? 3 is to renew this timer?
So the design of newcommandtimeout is for a session that doesn't operate for too long?

**Comment by mykola-mokhnach** [comment]
Maybe the documentation is not precise enough, but this is how the capability was always working. It only deletes the session when the next session command arrives after the timeout.

**Comment by mykola-mokhnach** [comment]
I went through the code again. It still looks to me like it was always working like this (e.g. the timeout was only reset after successful command execution). Also your comments also convince me the timer should be reset after command failure, because the session is still active.

I've prepared the PR https://github.com/appium/appium/pull/20379 to fix it, although it needs more thorough testing.  I would appreciate it if someone could help me with that
